base/collection_xattr.go

```diff
@@ -428,68 +452,105 @@ func (c *Collection) SubdocSetXattrs(k string, xvs map[string][]byte) (casOut ui
+		}
 		return 0, mutateErr
 	}
 	return uint64(result.Cas()), nil
 }
 
 // updateBodyAndXattr updates the document body and xattrs of an existing document. Writes cas and crc32c to the xattr using macro expansion.
-func (c *Collection) updateBodyAndXattrs(ctx context.Context, k string, exp uint32, cas uint64, opts *sgbucket.MutateInOptions, v interface{}, xattrs map[string][]byte) (casOut uint64, err error) {
+func (c *Collection) updateBodyAndXattrs(ctx context.Context, k string, exp uint32, cas uint64, opts *sgbucket.MutateInOptions, v interface{}, xattrs map[string][]byte, xattrsToDelete []string) (casOut uint64, err error) {
 	c.Bucket.waitForAvailKvOp()
 	defer c.Bucket.releaseKvOp()
 
-	mutateOps := make([]gocb.MutateInSpec, 0, len(xattrs)+1)
-	for xattrKey, xattrVal := range xattrs {
-		mutateOps = append(mutateOps, gocb.UpsertSpec(xattrKey, bytesToRawMessage(xattrVal), UpsertSpecXattr))
+	mutateOps, err := getUpsertSpecs(xattrs)
+	if err != nil {
+		return 0, err
+	}
+	for _, xattrKey := range xattrsToDelete {
+		if _, ok := xattrs[xattrKey]; ok {
+			return 0, fmt.Errorf("%s: %w", xattrKey, sgbucket.ErrUpsertAndDeleteSameXattr)
+		}
+		mutateOps = append(mutateOps, gocb.RemoveSpec(xattrKey, RemoveSpecXattr))
+
 	}
 	mutateOps = append(mutateOps, gocb.ReplaceSpec("", bytesToRawMessage(v), nil))
 	mutateOps = appendMacroExpansions(mutateOps, opts)
 
 	options := &gocb.MutateInOptions{
 		Expiry:        CbsExpiryToDuration(exp),
-		StoreSemantic: gocb.StoreSemanticsUpsert,
+		StoreSemantic: gocb.StoreSemanticsReplace,
 		Cas:           gocb.Cas(cas),
 	}
 	fillMutateInOptions(ctx, options, opts)
 	result, mutateErr := c.Collection.MutateIn(k, mutateOps, options)
 	if mutateErr != nil {
+		if errors.Is(mutateErr, gocb.ErrDocumentNotFound) {
+			casMismatchErr := sgbucket.CasMismatchErr{Actual: 0, Expected: cas}
+			mutateErr = fmt.Errorf("%w, gocb err: %w", casMismatchErr, mutateErr)
+		}
 		return 0, mutateErr
 	}
 	return uint64(result.Cas()), nil
 }
 
 // updateXattrDeleteBody deletes the document body and updates the xattrs of an existing document. Writes cas and crc32c to the xattr using macro expansion.
-func (c *Collection) updateXattrsDeleteBody(_ context.Context, k string, exp uint32, cas uint64, xattrs map[string][]byte, opts *sgbucket.MutateInOptions) (casOut uint64, err error) {
+func (c *Collection) updateXattrsDeleteBody(_ context.Context, k string, exp uint32, cas uint64, xattrs map[string][]byte, xattrsToDelete []string, opts *sgbucket.MutateInOptions) (casOut uint64, err error) {
 	c.Bucket.waitForAvailKvOp()
 	defer c.Bucket.releaseKvOp()
 
-	mutateOps := make([]gocb.MutateInSpec, 0, len(xattrs)+1)
-	for xattrKey, xattrVal := range xattrs {
-		mutateOps = append(mutateOps, gocb.UpsertSpec(xattrKey, bytesToRawMessage(xattrVal), UpsertSpecXattr))
+	mutateOps, err := getUpsertSpecs(xattrs)
+	if err != nil {
+		return 0, err
+	}
+	if cas == 0 && len(xattrsToDelete) > 0 {
+		return 0, sgbucket.ErrDeleteXattrOnDocumentInsert
+	}
```

Comment:
We don't do this check in other methods that attempt to delete xattrs (like updateXattrs).  Is it needed there also?

Reply:
`updateXattrs` is a private function and this is checked in the callers to this function in `WriteTombstoneWithXattrs` and `WriteWithXattrs`. `UpdateXattrs` doesn't allow you to delete xattrs since that's not yet needed. I think this has the potential to go wrong if the code is refactored and I like the idea to put this check in `updateXattrs` as well, though it will not presently be hit.